File: composeApp/src/androidMain/kotlin/com/muralex/exploramus/navigation/ScreenPicker.kt
package com.muralex.exploramus.navigation

import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import com.muralex.exploramus.screens.details.DetailsScreen
import com.muralex.exploramus.screens.favorites.FavoritesScreen
import com.muralex.exploramus.screens.home.HomeScreen
import com.muralex.exploramus.screens.search.SearchScreen
import com.muralex.exploramus.screens.section.SectionScreen
import com.muralex.exploramus.screens.settings.SettingsScreen
import com.muralex.exploramus.viewmodel.core.Navigation
import com.muralex.exploramus.viewmodel.core.ScreenIdentifier
import com.muralex.exploramus.viewmodel.screens.Screen
import com.muralex.exploramus.viewmodel.screens.countrydetail.DetailsScreenState
import com.muralex.exploramus.viewmodel.screens.favorites.FavoritesScreenState
import com.muralex.exploramus.viewmodel.screens.home.HomeScreenState
import com.muralex.exploramus.viewmodel.screens.search.SearchScreenState
import com.muralex.exploramus.viewmodel.screens.section.SectionScreenState
import com.muralex.exploramus.viewmodel.screens.settings.SettingsScreenState

@Composable
fun Navigation.ScreenPicker(
    screenIdentifier: ScreenIdentifier,
    screenNavActions: ScreenNavActions,
) {
    val state by stateProvider.getScreenStateFlow(screenIdentifier).collectAsState()

    val eventHandlers = remember {
        EventHandlers(events = events, navActions = screenNavActions)
    }

    when (screenIdentifier.screen) {
        Screen.HomeScreen ->
            HomeScreen(
                screenState = state as HomeScreenState,
                eventHandler = eventHandlers.home,
            )

        Screen.FavoritesScreen ->
            FavoritesScreen(
                screenState = state as FavoritesScreenState,
                eventHandler = eventHandlers.favorites,
            )

        Screen.SectionScreen ->
            SectionScreen(
                screenState = state as SectionScreenState,
                eventHandler = eventHandlers.section,
            )
        Screen.CountryDetail ->
            DetailsScreen(
                screenState = state as DetailsScreenState,
                eventHandler = eventHandlers.details,
            )

        Screen.SearchScreen ->
            SearchScreen(
                screenState = state as SearchScreenState,
                eventHandler = eventHandlers.search,
            )

        Screen.SettingsScreen ->
            SettingsScreen(
                screenState = state as SettingsScreenState,
                eventHandler = eventHandlers.settings,
            )

        else -> {}
    }

}

File: iosApp/iosApp/views/navigation/ScreenPicker.swift
import SwiftUI
import Shared

struct ScreenPicker: View {
    let requestedSId: ScreenIdentifier
    let level1ScreenIdentifier: ScreenIdentifier
    
    @EnvironmentObject var appObj: AppObservableObject
    @Environment(\.appTheme) var theme
    @Environment(\.appLayout) var appLayout
    
    @StateObject private var eventHandlers: EventHandlers
    
    init(requestedSId: ScreenIdentifier, level1ScreenIdentifier: ScreenIdentifier, eventHandlers: EventHandlers) {
        self.requestedSId = requestedSId
        self.level1ScreenIdentifier = level1ScreenIdentifier
        _eventHandlers = StateObject(wrappedValue: eventHandlers)
    }
    
    var body: some View {
        let state = appObj.getScreenState(sID: requestedSId)
        let navigation = appObj.dkmpNav
        let isLevel1 = requestedSId.screen.navigationLevel == 1
        
        ZStack {
            Color.clear.ignoresSafeArea()
            
            switch requestedSId.screen {
                
            case .homeScreen:
                HomeScreen(
                    screenState: state as! HomeScreenState,
                    eventHandler: eventHandlers.home
                )
                
            case .sectionScreen:
                SectionScreen(
                    screenState: state as! SectionScreenState,
                    eventHandler: eventHandlers.section
                )
                
                
            case .countryDetail:
                DetailsScreen(
                    screenState: state as! DetailsScreenState,
                    eventHandler: eventHandlers.details
                )
                
            case .favoritesScreen:
                FavoritesScreen(
                    screenState: state as! FavoritesScreenState,
                    eventHandler: eventHandlers.favorites
                )
                
            case .searchScreen:
                SearchScreen(
                    screenState: state as! SearchScreenState,
                    eventHandler: eventHandlers.search
                )
                
            default: EmptyView()
                
            }
        }
        .background(theme.background)
        .safeAreaInset(edge: .bottom) {
            if isLevel1 {
                Color.clear.frame(height: 80)
            }
        }
        .navigationTitle(navigation.getTitle(screenIdentifier: requestedSId))
        .navigationBarTitleDisplayMode(.inline)
        .onDisappear {
            navigation.exitScreenForIos(screenIdentifier: requestedSId)
        }
        .task {
            await appObj.collectScreenStateFlow(sID: requestedSId)
        }
        .toolbar {
            if isLevel1 && !appLayout.useDrawer {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button {
                        eventHandlers.navActions.showSettings()
                    } label: {
                        Image(systemName: "gearshape")
                    }
                }
            }
        }
    }
}





